README.md
# Typogruby

[![Build Status](https://secure.travis-ci.org/avdgaag/typogruby.png?branch=master)](http://travis-ci.org/avdgaag/typogruby)

Helps you improve your web typograpbhy with some standard text filters.

This project is based on Django's Typogrify, so the best introduction to read would be [Jeff Croft's][1].

I created this gem to easily share these text filters in some tiny Ruby projects, including [a TextMate bundle][5]. For production code I recommend checking out the originals first.

## General Usage

First, install the Ruby gem:

    $ gem install typogruby

Then require the library to get started:

```ruby
require 'typogruby'
Typogruby.improve('Hello, world!')
```

Or, you can include the library in a helper or something:

```ruby
require 'typogruby'
include Typogruby
improve('Hello, world!')
```

See the [full API documentation][4] for more information. You could also use [this Textmate Bundle][5].

## From the command line

You can also use typogruby directly from the command line:

    typogruby MY_FILE

This will output the contents of your file with all filters applied. Use `typogruby -h` for more information and options.

## References

* Based on [typography-helper][2]
* ...and on [Typogrify][3]
* [Description of typogrify][1]

## Changelog

See HISTORY.md for the complete changelog.

## Note on Patches/Pull Requests

* Fork the project.
* Make your feature addition or bug fix.
* Add tests for it. This is important so I don't break it in a
  future version unintentionally.
* Commit, do not mess with rakefile, version, or history.
  (if you want to have your own version, that is fine but bump version in a commit by itself I can ignore when I pull)
* Send me a pull request. Bonus points for topic branches.

## Credits

By Arjan van der Gaag ([avdgaag @ github][6]) based on the hard work of lots of others. See LICENSE for license details (same as Ruby).

This gem includes contributions by:

* Justin Hileman
* Peter Aronoff

[1]: http://jeffcroft.com/blog/2007/may/29/typogrify-easily-produce-web-typography-doesnt-suc/
[2]: http://github.com/hunter/typography-helper
[3]: http://code.google.com/p/typogrify
[4]: http://avdgaag.github.com/typogruby
[5]: http://github.com/avdgaag/Typography-tmbundle
[6]: http://github.com/avdgaag


HISTORY.md
# Changelog

## 1.0.18

* Added `<title>` to list of excluded tags
* Switched to Minitest for unit testing

## 1.0.17

* Allow ampersands in capitalised words (thanks Paul Robert Lloyd)

## 1.0.16

* Added textarea to list of ignored elements (thanks Juan Antonio Navarro Pérez)
* Added Haml filter

## 1.0.15

* Added head to list of ignored elements (thanks Peter Aronoff)

## 1.0.14

* Bugfix: ignore multi-line HTML attributes in widont filter

## 1.0.13

* Be smart about periods in abbreviations. Capitalise LA, L.A. and L.A, but not L. or L. A. (by Brad Weslake)

## 1.0.12

* Fixed aruba deprecation warnings
* Ensure duplicate sensitive tags are properly rendered (thanks Justin Hileman)
* Removed obsolete tag-ignoring clauses from regexes

## 1.0.11

* Excluded more typography-sensitive tags from filtering (thanks Justin Hileman)

## 1.0.10

* Decoupled jeweler, which is no longer a dependency.
* Various minor documentation issues and internal refactorings

## 1.0.9

* Bugfix: ignore caps in HTML attribute values

## 1.0.8

* Allow command-line output to disk via -o option
* command-line tool reads both files and STDIN, enabling piping of content
* added filter to replace special characters with HTML entities
* better tests for the command-line tool with cucumber and aruba

## 1.0.7

* Improved Ruby 1.9 compatibility
* Improved documentation
* Merged test helper and test case

## 1.0.6

* Bugfix: ignore inline javascript

## 1.0.5

* Manage gem dependencies using Bundler

## 1.0.4

* Bugfix: no longer regard combination of digits and periods as caps.
* Bugfix: do not add consecutive non-breaking spaces to prevent widows

## 1.0.3

* Bugfix: caps also ignores unequal but same excepted tags
* Feature: added command-line program
* Improved documentation

## 1.0.2

* Bugfix: initial_quotes also wraps named HTML entities
* Bugfix: caps no longer wraps consecutive numbers

## 1.0

* First release


test/helper.rb
require 'rubygems'
require 'bundler'
require 'test/unit'

Bundler.setup

$LOAD_PATH.unshift(File.dirname(__FILE__))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
require 'typogruby'

class Test::Unit::TestCase
  include Typogruby
end


test/test_typogrify.rb
require 'helper'

class TestTypogrify < Test::Unit::TestCase

  def test_should_replace_amps
    assert_equal 'One <span class="amp">&amp;</span> two', amp('One & two')
    assert_equal 'One <span class="amp">&amp;</span> two', amp('One &amp; two')
    assert_equal 'One <span class="amp">&amp;</span> two', amp('One &#38; two')
    assert_equal 'One&nbsp;<span class="amp">&amp;</span>&nbsp;two', amp('One&nbsp;&amp;&nbsp;two')
  end

  def test_should_ignore_special_amps
    assert_equal 'One <span class="amp">&amp;</span> two', amp('One <span class="amp">&amp;</span> two')
    assert_equal '&ldquo;this&rdquo; <span class="amp">&amp;</span> <a href="/?that&amp;test">that</a>', amp('&ldquo;this&rdquo; & <a href="/?that&amp;test">that</a>')
  end

  def test_should_ignore_standalone_amps_in_attributes
    assert_equal '<link href="xyz.html" title="One &amp; Two">xyz</link>', amp('<link href="xyz.html" title="One & Two">xyz</link>')
  end

  def test_should_replace_caps
    assert_equal 'A message from <span class="caps">KU</span>', caps("A message from KU")
  end

  def test_should_ignore_special_case_caps
    assert_equal '<pre>CAPS</pre> more <span class="caps">CAPS</span>', caps("<pre>CAPS</pre> more CAPS")
    assert_equal 'A message from <span class="caps">2KU2</span> with digits', caps("A message from 2KU2 with digits")
    assert_equal 'Dotted caps followed by spaces should never include them in the wrap <span class="caps">D.O.T.</span>   like so.', caps("Dotted caps followed by spaces should never include them in the wrap D.O.T.   like so.")
  end

  def test_should_not_break_caps_with_apostrophes
    assert_equal '<span class="caps">JIMMY\'S</span>', caps("JIMMY'S")
    assert_equal '<i><span class="caps">D.O.T.</span></i><span class="caps">HE34T</span><b><span class="caps">RFID</span></b>', caps("<i>D.O.T.</i>HE34T<b>RFID</b>")
  end

  def test_should_replace_quotes
    assert_equal '<span class="dquo">"</span>With primes"', initial_quotes('"With primes"')
    assert_equal '<span class="quo">\'</span>With single primes\'', initial_quotes("'With single primes'")
    assert_equal '<a href="#"><span class="dquo">"</span>With primes and a link"</a>', initial_quotes('<a href="#">"With primes and a link"</a>')
    assert_equal '<span class="dquo">&#8220;</span>With smartypanted quotes&#8221;', initial_quotes('&#8220;With smartypanted quotes&#8221;')
  end

  def test_should_apply_smartypants
    assert_equal 'The &#8220;Green&#8221; man', smartypants('The "Green" man')
  end

  def test_should_apply_all_filters
    assert_equal '<h2><span class="dquo">&#8220;</span>Jayhawks&#8221; <span class="amp">&amp;</span> <span class="caps">KU</span> fans act extremely&nbsp;obnoxiously</h2>', improve('<h2>"Jayhawks" & KU fans act extremely obnoxiously</h2>')
  end

  def test_should_prevent_widows
    assert_equal 'A very simple&nbsp;test', widont('A very simple test')
  end

  def test_should_not_change_single_word_items
    assert_equal 'Test', widont('Test')
    assert_equal ' Test', widont(' Test')
    assert_equal '<ul><li>Test</p></li><ul>', widont('<ul><li>Test</p></li><ul>')
    assert_equal '<ul><li> Test</p></li><ul>', widont('<ul><li> Test</p></li><ul>')
    assert_equal '<p>In a couple of&nbsp;paragraphs</p><p>paragraph&nbsp;two</p>', widont('<p>In a couple of paragraphs</p><p>paragraph two</p>')
    assert_equal '<h1><a href="#">In a link inside a&nbsp;heading</i> </a></h1>', widont('<h1><a href="#">In a link inside a heading</i> </a></h1>')
    assert_equal '<h1><a href="#">In a link</a> followed by other&nbsp;text</h1>', widont('<h1><a href="#">In a link</a> followed by other text</h1>')
  end

  def test_should_not_error_on_empty_html
    assert_equal '<h1><a href="#"></a></h1>', widont('<h1><a href="#"></a></h1>')
  end

  def test_should_ignore_widows_in_special_tags
    assert_equal '<div>Divs get no love!</div>', widont('<div>Divs get no love!</div>')
    assert_equal '<pre>Neither do PREs</pre>', widont('<pre>Neither do PREs</pre>')
    assert_equal '<div><p>But divs with paragraphs&nbsp;do!</p></div>', widont('<div><p>But divs with paragraphs do!</p></div>')
  end
end

test/test_typogruby.rb
# -*- encoding: utf-8 -*-
require 'minitest/autorun'
require File.expand_path('../../lib/typogruby', __FILE__)

class TypogrubyTest < Minitest::Test
  include Typogruby

  def test_should_replace_amps
    assert_equal 'One <span class="amp">&amp;</span> two', amp('One & two')
    assert_equal 'One <span class="amp">&amp;</span> two', amp('One &amp; two')
    assert_equal 'One <span class="amp">&amp;</span> two', amp('One &#38; two')
    assert_equal 'One&nbsp;<span class="amp">&amp;</span>&nbsp;two', amp('One&nbsp;&amp;&nbsp;two')
  end

  def test_should_ignore_special_amps
    assert_equal 'One <span class="amp">&amp;</span> two', amp('One <span class="amp">&amp;</span> two')
    assert_equal '&ldquo;this&rdquo; <span class="amp">&amp;</span> <a href="/?that&amp;test">that</a>', amp('&ldquo;this&rdquo; & <a href="/?that&amp;test">that</a>')
    assert_equal %Q{<script>\nvar x = "FOO & BAR";\n</script>"}, initial_quotes(%Q{<script>\nvar x = "FOO & BAR";\n</script>"})
  end

  def test_should_ignore_standalone_amps_in_attributes
    assert_equal '<link href="xyz.html" title="One &amp; Two">xyz</link>', amp('<link href="xyz.html" title="One & Two">xyz</link>')
  end

  def test_should_replace_caps
    assert_equal 'A message from <span class="caps">KU</span>', caps("A message from KU")
    assert_equal 'Replace text <a href="."><span class="caps">IN</span></a> tags', caps('Replace text <a href=".">IN</a> tags')
    assert_equal 'Replace text <i><span class="caps">IN</span></i> tags', caps('Replace text <i>IN</i> tags')
  end

  def test_should_ignore_special_case_caps
    assert_equal 'It should ignore just numbers like 1234.', caps('It should ignore just numbers like 1234.')
    assert_equal '<pre>CAPS</pre> more <span class="caps">CAPS</span>', caps("<pre>CAPS</pre> more CAPS")
    assert_equal '<Pre>CAPS</PRE> with odd tag names <span class="caps">CAPS</span>', caps("<Pre>CAPS</PRE> with odd tag names CAPS")
    assert_equal 'A message from <span class="caps">2KU2</span> with digits', caps("A message from 2KU2 with digits")
    assert_equal 'Dotted caps followed by spaces should never include them in the wrap <span class="caps">D.O.T.</span>   like so.', caps("Dotted caps followed by spaces should never include them in the wrap D.O.T.   like so.")
    assert_equal 'Caps in attributes (<span title="Example CAPS">example</span>) should be ignored', caps('Caps in attributes (<span title="Example CAPS">example</span>) should be ignored')
    assert_equal '<head><title>CAPS Example</title></head>', caps('<head><title>CAPS Example</title></head>')
  end

  def test_should_not_break_caps_with_apostrophes
    assert_equal '<span class="caps">JIMMY\'S</span>', caps("JIMMY'S")
    assert_equal '<i><span class="caps">D.O.T.</span></i><span class="caps">HE34T</span><b><span class="caps">RFID</span></b>', caps("<i>D.O.T.</i>HE34T<b>RFID</b>")
  end

  def test_should_not_break_caps_with_ampersands
    assert_equal '<span class="caps">AT&T</span>', caps("AT&T")
    assert_equal '<span class="caps">AT&amp;T</span>', caps("AT&amp;T")
    assert_equal '<span class="caps">AT&#38;T</span>', caps("AT&#38;T")
  end

  def test_should_replace_quotes
    assert_equal '<span class="dquo">"</span>With primes"', initial_quotes('"With primes"')
    assert_equal '<span class="quo">\'</span>With single primes\'', initial_quotes("'With single primes'")
    assert_equal '<a href="#"><span class="dquo">"</span>With primes and a link"</a>', initial_quotes('<a href="#">"With primes and a link"</a>')
    assert_equal '<span class="dquo">&#8220;</span>With smartypanted quotes&#8221;', initial_quotes('&#8220;With smartypanted quotes&#8221;')
    assert_equal '<span class="quo">&lsquo;</span>With manual quotes&rsquo;', initial_quotes('&lsquo;With manual quotes&rsquo;')
    assert_equal %Q{<script>\nvar x = "FOO BAR and BAZ";\n</script>"}, initial_quotes(%Q{<script>\nvar x = "FOO BAR and BAZ";\n</script>"})
  end

  def test_should_not_replace_quotes_in_scripts
    assert_equal %Q{<script>\nvar x = "foo" +\n"bar";\n</script>"}, initial_quotes(%Q{<script>\nvar x = "foo" +\n"bar";\n</script>"})
  end

  def test_should_apply_smartypants
    assert_equal 'The &#8220;Green&#8221; man', smartypants('The "Green" man')
  end

  def test_should_apply_all_filters
    assert_equal '<h2><span class="dquo">&#8220;</span>Jayhawks&#8221; <span class="amp">&amp;</span> <span class="caps">KU</span> fans act extremely&nbsp;obnoxiously</h2>', improve('<h2>"Jayhawks" & KU fans act extremely obnoxiously</h2>')
  end

  def test_should_prevent_widows
    assert_equal 'A very simple&nbsp;test', widont('A very simple test')
  end

  def test_should_not_change_single_word_items
    assert_equal 'Test', widont('Test')
    assert_equal ' Test', widont(' Test')
    assert_equal '<ul><li>Test</p></li><ul>', widont('<ul><li>Test</p></li><ul>')
    assert_equal '<ul><li> Test</p></li><ul>', widont('<ul><li> Test</p></li><ul>')
    assert_equal '<p>In a couple of&nbsp;paragraphs</p><p>paragraph&nbsp;two</p>', widont('<p>In a couple of paragraphs</p><p>paragraph two</p>')
    assert_equal '<h1><a href="#">In a link inside a&nbsp;heading</i> </a></h1>', widont('<h1><a href="#">In a link inside a heading</i> </a></h1>')
    assert_equal '<h1><a href="#">In a link</a> followed by other&nbsp;text</h1>', widont('<h1><a href="#">In a link</a> followed by other text</h1>')
  end

  def test_should_not_add_nbsp_before_another
    assert_equal 'Sentence with one&nbsp;nbsp', widont('Sentence with one&nbsp;nbsp')
  end

  def test_should_not_error_on_empty_html
    assert_equal '<h1><a href="#"></a></h1>', widont('<h1><a href="#"></a></h1>')
  end

  def test_should_ignore_caps_in_special_tags
    assert_equal '<title>I like JSON</title>', caps('<title>I like JSON</title>')
  end

  def test_should_ignore_widows_in_special_tags
    assert_equal '<div>Divs get no love!</div>', widont('<div>Divs get no love!</div>')
    assert_equal '<pre>Neither do PREs</pre>', widont('<pre>Neither do PREs</pre>')
    assert_equal '<title>or titles</title>', widont('<title>or titles</title>')
    assert_equal '<textarea>nor text in textarea</textarea>', widont('<textarea>nor text in textarea</textarea>')
    assert_equal "<script>\nreturn window;\n</script>", widont("<script>\nreturn window;\n</script>")
    assert_equal '<div><p>But divs with paragraphs&nbsp;do!</p></div>', widont('<div><p>But divs with paragraphs do!</p></div>')
  end

  def test_should_ignore_attributes_across_mutliple_lines_when_preventing_widows
    assert_equal %Q{<div id="aniver" style="\nbackground-image:url('/static/colorfulbackground.png');\nbackground-repeat:no-repeat;\ntext-align:center;\nmargin:0;\npadding:0;\nbackground-size: 100% 100%;">},
      widont(%Q{<div id="aniver" style="\nbackground-image:url('/static/colorfulbackground.png');\nbackground-repeat:no-repeat;\ntext-align:center;\nmargin:0;\npadding:0;\nbackground-size: 100% 100%;">})
  end

  def test_should_convert_entities
    assert_equal "Vari&euml;ren&hellip;", entities('Variëren…')
    assert_equal "<p>Ol&eacute;</p>", entities('<p>Olé</p>')
  end

  def test_should_leave_sensitive_tags_alone
    %w{script pre code kbd math}.each do |tag_name|
      test_string = "<#{tag_name}>“CAPS & chårs” and widows</#{tag_name}>"
      assert_equal test_string, improve(test_string)
    end
  end

  def test_should_not_forget_about_duplicate_sensitive_tags
    %w{script pre code kbd math}.each do |tag_name|
      test_string = "<#{tag_name}>this</#{tag_name}>==<#{tag_name}>this</#{tag_name}>"
      assert_equal test_string, improve(test_string)
    end
  end
end


lib/typogruby.rb
# -*- encoding: utf-8 -*-

require 'rubypants'
require 'digest/md5'
$KCODE = 'U' if RUBY_VERSION < '1.9.0'

# A collection of simple helpers for improving web
# typograhy. Based on TypographyHelper by Luke Hartman and Typogrify.
#
# @example Using all filters
#   require 'typogruby'
#   Typogruby.improve('my text')
#
# @example Using a single filter
#   Typogruby.initial_quotes('my text')
#
# @see http://github.com/hunter/typography-helper
# @see http://code.google.com/p/typogrify
# @author Arjan van der Gaag <arjan.vandergaag@gmail.com>
module Typogruby

  # Applies smartypants to a given piece of text
  #
  # @example
  #   smartypants('The "Green" man')
  #   # => 'The &#8220;Green&#8221; man'
  #
  # @see https://rubygems.org/gems/rubypants
  # @param [String] text input text
  # @return [String] input text with smartypants applied
  def smartypants(text)
    ::RubyPants.new(text).to_html
  end

  # converts a & surrounded by optional whitespace or a non-breaking space
  # to the HTML entity and surrounds it in a span with a styled class.
  #
  # @example
  #   amp('One & two')
  #   # => 'One <span class="amp">&amp;</span> two'
  #   amp('One &amp; two')
  #   # => 'One <span class="amp">&amp;</span> two'
  #   amp('One &#38; two')
  #   # => 'One <span class="amp">&amp;</span> two'
  #   amp('One&nbsp;&amp;&nbsp;two')
  #   # => 'One&nbsp;<span class="amp">&amp;</span>&nbsp;two'
  #
  # @example It won't mess up & that are already wrapped, in entities or URLs
  #   amp('One <span class="amp">&amp;</span> two')
  #   # => 'One <span class="amp">&amp;</span> two'
  #   amp('&ldquo;this&rdquo; & <a href="/?that&amp;test">that</a>')
  #   # => '&ldquo;this&rdquo; <span class="amp">&amp;</span> <a href="/?that&amp;test">that</a>'
  #
  # @example It should ignore standalone amps that are in attributes
  #   amp('<link href="xyz.html" title="One & Two">xyz</link>')
  #   # => '<link href="xyz.html" title="One & Two">xyz</link>'
  #
  # @param [String] text input text
  # @return [String] input text with ampersands wrapped
  def amp(text)
    # $1 is the part before the caps and $2 is the amp match
    exclude_sensitive_tags(text) do |t|
      t.gsub(/(\s|&nbsp;)&(?:amp;|#38;)?(\s|&nbsp;)/) { |str|
        $1 + '<span class="amp">&amp;</span>' + $2
      }.gsub(/(\w+)="(.*?)<span class="amp">&amp;<\/span>(.*?)"/, '\1="\2&amp;\3"')
    end
  end

  # replaces space(s) before the last word (or tag before the last word)
  # before an optional closing element (<tt>a</tt>, <tt>em</tt>,
  # <tt>span</tt>, strong) before a closing tag (<tt>p</tt>, <tt>h[1-6]</tt>,
  # <tt>li</tt>, <tt>dt</tt>, <tt>dd</tt>) or the end of the string.
  #
  # @example
  #   widont('A very simple test')
  #   # => 'A very simple&nbsp;test'
  #
  # @example Single word items shouldn't be changed
  #   widont('Test')
  #   # => 'Test'
  #   widont(' Test')
  #   # => ' Test'
  #   widont('<ul><li>Test</p></li><ul>')
  #   # => '<ul><li>Test</p></li><ul>'
  #   widont('<ul><li> Test</p></li><ul>')
  #   # => '<ul><li> Test</p></li><ul>'
  #
  # @example Nested tags
  #   widont('<p>In a couple of paragraphs</p><p>paragraph two</p>')
  #   # => '<p>In a couple of&nbsp;paragraphs</p><p>paragraph&nbsp;two</p>'
  #   widont('<h1><a href="#">In a link inside a heading</i> </a></h1>')
  #   # => '<h1><a href="#">In a link inside a&nbsp;heading</i> </a></h1>'
  #   widont('<h1><a href="#">In a link</a> followed by other text</h1>')
  #   # => '<h1><a href="#">In a link</a> followed by other&nbsp;text</h1>'
  #
  # @example Empty HTMLs shouldn't error
  #   widont('<h1><a href="#"></a></h1>')
  #   # => '<h1><a href="#"></a></h1>'
  #
  # @example Excluded tags
  #   widont('<div>Divs get no love!</div>')
  #   # => '<div>Divs get no love!</div>'
  #   widont('<pre>Neither do PREs</pre>')
  #   # => '<pre>Neither do PREs</pre>'
  #   widont('<div><p>But divs with paragraphs do!</p></div>')
  #   # => '<div><p>But divs with paragraphs&nbsp;do!</p></div>'
  #
  # @see http://mucur.name/posts/widon-t-and-smartypants-helpers-for-rails
  # @see http://shauninman.com/archive/2006/08/22/widont_wordpress_plugin
  # @param [String] text input text
  # @return [String] input text with non-breaking spaces inserted
  def widont(text)
    exclude_sensitive_tags(text) do |t|
      t.gsub(%r{
        (<[^/][^>]*?>)|                                 # Ignore any opening tag, so we don't mess up attribute values
        ((?:</?(?:a|em|span|strong|i|b)[^>]*>)|[^<>\s]) # must be proceeded by an approved inline opening or closing tag or a nontag/nonspace
        \s+                                             # the space to replace
        (([^<>\s]+)                                     # must be flollowed by non-tag non-space characters
        \s*                                             # optional white space!
        (</(a|em|span|strong|i|b)>\s*)*                 # optional closing inline tags with optional white space after each
        ((</(p|h[1-6]|li|dt|dd)>)|$))                   # end with a closing p, h1-6, li or the end of the string
      }xm) { |match| $1 ? match : $2 + (match.include?('&nbsp;') ? ' ' : '&nbsp;') + $3 } # Make sure to not add another nbsp before one already there
    end
  end

  # surrounds two or more consecutive captial letters, perhaps with interspersed digits and periods
  # in a span with a styled class.
  #
  # @example
  #   caps("A message from KU")
  #   # => 'A message from <span class="caps">KU</span>'
  #
  # @example Allows digits
  #   caps("A message from 2KU2 with digits")
  #   # => 'A message from <span class="caps">2KU2</span> with digits'
  #
  # @example Allows ampersands
  #   caps("A phone bill from AT&T")
  #   # => 'A phone bill from <span class="caps">AT&amp;T</span>'
  #
  # @example Ignores HTML attributes
  #   caps('Download <a href="file.doc" title="PDF document">this file</a>')
  #   # => 'Download <a href="file.doc" title="PDF document">this file</a>'
  #
  # @example All caps with with apostrophes in them shouldn't break. Only handles dump apostrophes though.
  #   caps("JIMMY'S")
  #   # => '<span class="caps">JIMMY\\'S</span>'
  #   caps("<i>D.O.T.</i>HE34T<b>RFID</b>")
  #   # => '<i><span class="caps">D.O.T.</span></i><span class="caps">HE34T</span><b><span class="caps">RFID</span></b>'
  #
  # @example Sequences of capitalised letters with spaces are not eligible, since they may be names.
  #   caps("L.A.P.D")
  #   # => '<span class="caps">L.A.P.D.</span>'
  #   caps("L. A. Paul")
  #   # => 'L. A. Paul'
  #
  # @param [String] text input text
  # @return [String] input text with caps wrapped
  def caps(text)
    exclude_sensitive_tags(text) do |t|
      # $1 and $2 are excluded HTML tags, $3 is the part before the caps and $4 is the caps match
      t.gsub(%r{
          (<[^/][^>]*?>)|                                      # Ignore any opening tag, so we don't mess up attribute values
          (\s|&nbsp;|^|'|"|>|)                                 # Make sure our capture is preceded by whitespace or quotes
          ([A-Z\d](?:(\.|'|&|&amp;|&\#38;)?[A-Z\d][\.']?){1,}) # Capture capital words, with optional dots, numbers or ampersands in between
          (?!\w)                                               # ...which must not be followed by a word character.
        }x) do |str|
        tag, before, caps = $1, $2, $3

        # Do nothing with the contents if ignored tags, the inside of an opening HTML element
        # so we don't mess up attribute values, or if our capture is only digits.
        if tag || caps =~ /^\d+\.?$/
          str
        elsif $3 =~ /^[\d\.]+$/
          before + caps
        else
          before + '<span class="caps">' + caps + '</span>'
        end
      end
    end
  end

  # encloses initial single or double quote, or their entities
  # (optionally preceeded by a block element and perhaps an inline element)
  # with a span that can be styled.
  #
  # @example
  #   initial_quotes('"With primes"')
  #   # => '<span class="dquo">"</span>With primes"'
  #   initial_quotes("'With single primes'")
  #   # => '<span class="quo">\\'</span>With single primes\\''
  #
  # @example With primes and links
  #   initial_quotes('<a href="#">"With primes and a link"</a>')
  #   # => '<a href="#"><span class="dquo">"</span>With primes and a link"</a>'
  #
  # @example with Smartypants-quotes
  #   initial_quotes('&#8220;With smartypanted quotes&#8221;')
  #   # => '<span class="dquo">&#8220;</span>With smartypanted quotes&#8221;'
  #
  # @param [String] text input text
  # @return [String] input text with initial quotes wrapped
  def initial_quotes(text)
    # $1 is the initial part of the string, $2 is the quote or entitity, and $3 is the double quote
    exclude_sensitive_tags(text) do |t|
      t.gsub(/((?:<(?:h[1-6]|p|li|dt|dd)[^>]*>|^)\s*(?:<(?:a|em|strong|span)[^>]*>)?)('|&#8216;|&lsquo;|("|&#8220;|&ldquo;))/) {$1 + "<span class=\"#{'d' if $3}quo\">#{$2}</span>"}
    end
  end

  # Converts special characters (excluding HTML tags) to HTML entities.
  #
  # @example
  #   entities("Aloë Vera") # => "Alo&euml; Vera"
  #
  # @param [String] text input text
  # @return [String] input text with all special characters converted to
  #   HTML entities.
  def entities(text)
    o = ''
    text.scan(/(?x)

        ( <\?(?:[^?]*|\?(?!>))*\?>
        | <!-- (?m:.*?) -->
        | <\/? (?i:a|abbr|acronym|address|applet|area|b|base|basefont|bdo|big|blockquote|body|br|button|caption|center|cite|code|col|colgroup|dd|del|dfn|dir|div|dl|dt|em|fieldset|font|form|frame|frameset|h1|h2|h3|h4|h5|h6|head|hr|html|i|iframe|img|input|ins|isindex|kbd|label|legend|li|link|map|menu|meta|noframes|noscript|object|ol|optgroup|option|p|param|pre|q|s|samp|script|select|small|span|strike|strong|style|sub|sup|table|tbody|td|textarea|tfoot|th|thead|title|tr|tt|u|ul|var)\b
            (?:[^>"']|"[^"]*"|'[^']*')*
          >
        | &(?:[a-zA-Z0-9]+|\#[0-9]+|\#x[0-9a-fA-F]+);
        )
        |([^<&]+|[<&])

      /x) do |tag, text|
      o << tag.to_s
      o << encode(text.to_s)
    end
    o
  end

  # main function to do all the functions from the method.
  #
  # @param [String] text input text
  # @return [String] input text with all filters applied
  def improve(text)
    initial_quotes(caps(smartypants(widont(amp(text)))))
  end

private

  # Convert characters from the map in ./lib/characters.txt
  # Code taken from TextMate HTML bundle
  #
  # @param [String] text input text
  # @return [String] input text with all special characters converted to
  #   HTML entities.
  def encode(text)
    @char_to_entity ||= begin
      map = {}
      File.read(File.join(File.dirname(__FILE__), 'characters.txt')).scan(/^(\d+)\s*(.+)$/) do |key, value|
        map[[key.to_i].pack('U')] = value
      end
      map
    end

    text.gsub(/[^\x00-\x7F]|["'<>&]/) do |ch|
      ent = @char_to_entity[ch]
      ent ? "&#{ent};" : sprintf("&#x%02X;", ch.unpack("U")[0])
    end
  end

  # Hackish text filter that will make sure our text filters leave
  # sensitive tags alone without resorting to a full-blown HTML parser.
  #
  # Sensitive tags are tags with literal contents, which we do not
  # want to change. It currently ignores: <pre>, <code>, <kbd>, <math>
  # and <script>.
  #
  # The idea is simple: every text filter is applied as a block to this
  # method. This will preprocess the text and replace any sensitive tags
  # with a MD5 hash of its entire contents. Then the filter is called,
  # and then the hashes are replaced back with their original content.
  #
  # @yield [hashed_text] Hands you the input text with all sensitive tags
  #   hashed. The block's result will be unhashed and then returned.
  # @param [String] text
  # @return [String] input with sensitive tags restored
  def exclude_sensitive_tags(text)
    @exluded_sensitive_tags = {}
    modified_text = text.gsub(/<(#{EXCLUDED_TAGS .join('|')})[^>]*>.*?<\/\1>/mi) do |script|
      hash = Digest::MD5.hexdigest(script)
      @exluded_sensitive_tags[hash] = script
      hash
    end
    yield(modified_text).gsub(/#{@exluded_sensitive_tags.keys.join('|')}/) do |h|
      @exluded_sensitive_tags[h]
    end
  end

  # Array of all the senstive tags that should be ignored by all the text filters.
  EXCLUDED_TAGS  = %w{head pre code kbd math script textarea title}

  extend self
end
